### DISCUSSION TITLE
How to use cursor pagination within a public Table Builder?

### DISCUSSION THREAD

**Discussion by snarexx** [filamentphp/filament#10571] [Help]
I'm trying to use cursor navigation within a public table builder. 
In [this docs](https://filamentphp.com/docs/3.x/tables/advanced#using-cursor-pagination) it suggested that
```
You may use cursor pagination by overriding paginateTableQuery() method.

First, locate your Livewire component. If you're **using a resource from the Panel Builder and you want to add simple pagination to the List page**, you'll want to open the Pages/List.php file in the resource, not the resource class itself.

use Illuminate\Contracts\Pagination\CursorPaginator;
use Illuminate\Database\Eloquent\Builder;
 
protected function paginateTableQuery(Builder $query): CursorPaginator
{
    return $query->cursorPaginate(($this->getTableRecordsPerPage() === 'all') ? $query->count() : $this->getTableRecordsPerPage());
}
```
I'm not using panel builder for my public table builder, I have it for my admin panel. I don't have resource file for this. 
My question is how can I implement cursor pagination within a public Table Builder?
@danharrin  You recently committed this change. Maybe you may help me.
https://github.com/filamentphp/filament/commit/d55d952518c6ac17f10db5b07f5ad66de4309e67

**Comment by snarexx** [comment]
It worked when I changed to this

    protected function paginateTableQuery(Builder $query): Paginator | CursorPaginator
    {
        return $query->simplePaginate(($this->getTableRecordsPerPage() === 'all') ? $query->count() : $this->getTableRecordsPerPage());
    }

**Comment by cawecoy** [comment]
Hi @danharrin ! [Cursor pagination for Table Builder](https://filamentphp.com/docs/3.x/tables/advanced#using-cursor-pagination) is not working for me. The result is the same as [simple pagination](https://filamentphp.com/docs/3.x/tables/advanced#using-simple-pagination). Is the documentation missing something? Or is it really not working?

**Reply by danharrin** [reply]
The UI is the same as simple pagination as the technique does not calculate the number of pages available. However, the queries should be completely different to simple pagination.

**Reply by cawecoy** [reply]
OK, thanks @danharrin ! I need a pagination like this (see bellow). Is it possible with Filament?
![image](https://github.com/filamentphp/filament/assets/745305/63e1b0b3-8504-4122-83c9-3413d3f29087)

**Reply by danharrin** [reply]
Normal pagination?

**Reply by cawecoy** [reply]
@danharrin it's a little bit different:
<<   <   1  2  (3)  4  5   >    >>

Rather than the normal pagination which looks like:
<  1 2 (3) ... 22  23  24 >

Do you see that the pagination I need has no "..." in the middle (it actually shows only the two previous and the two next numbers in relation to the current page number) and has the "<<" and ">>" to go to the first and last page?

**Reply by danharrin** [reply]
How would you do that with normal Laravel pagination?

**Reply by cawecoy** [reply]
@danharrin copying [the Laravel pagination blade file](https://github.com/laravel/framework/blob/c4f46abce0e69737de35b71439943ac2a24d6ec3/src/Illuminate/Pagination/resources/views/default.blade.php#L3) and customizing it. This file is intuitive and has 47 lines of code, easy to customize.

Filament is meant to accelerate Laravel development, that's why I am using it. Anyway, [Filament pagination component](https://github.com/filamentphp/filament/blob/48390b8fc4953efdd0fd272a338d983ce4f994e8/packages/support/resources/views/components/pagination/index.blade.php#L8) has 167 lines of code which looks more complex, and [you don't recommend publishing/editing views](https://filamentphp.com/community/danharrin-publishing-views-in-laravel), that's why I am asking if there is an official solution that doesn't require touching the view files and that I am missing it.

It's not a matter of customizing CSS. I can't find anything like "<<" and ">>" to jump to first and last page. It's a design requirement at the project I am currently developing.

**Reply by danharrin** [reply]
The reason why it has more lines is because it has many more features and can handle multiple different types of pagination. Feel free to override this one if you need to, its not as risky as other views in the project.

**Reply by cawecoy** [reply]
Hi @danharrin ! I have created a PR to solve the issue: https://github.com/filamentphp/filament/pull/11278
